src/libpriv/rpmostree-core.c

```diff
@@ -915,18 +915,72 @@ rpmostree_context_download_metadata (RpmOstreeContext *self,
 {
   g_assert (!self->empty);
 
-  g_autoptr(DnfState) hifstate = dnf_state_new ();
+  g_auto(RpmSighandlerResetCleanup) rpmsigreset = { 0, };
 
-  guint progress_sigid = g_signal_connect (hifstate, "percentage-changed",
-                                           G_CALLBACK (on_hifstate_percentage_changed),
-                                           "Downloading metadata:");
+  g_autoptr(GPtrArray) rpmmd_repos = get_enabled_rpmmd_repos (self->hifctx, DNF_REPO_ENABLED_METADATA);
 
-  DECLARE_RPMSIGHANDLER_RESET;
-  if (!dnf_context_setup_sack (self->hifctx, hifstate, error))
-    return FALSE;
+  g_print ("Enabled rpm-md repositories:");
+  for (guint i = 0; i < rpmmd_repos->len; i++)
+    {
+      DnfRepo *repo = rpmmd_repos->pdata[i];
+      g_print (" %s", dnf_repo_get_id (repo));
+    }
+  g_print ("\n");
+
+  for (guint i = 0; i < rpmmd_repos->len; i++)
+    {
+      DnfRepo *repo = rpmmd_repos->pdata[i];
+      g_autoptr(DnfState) hifstate = dnf_state_new ();
+
+      gboolean did_update = FALSE;
+      if (!dnf_repo_check(repo,
+                          dnf_context_get_cache_age (self->hifctx),
+                          hifstate,
+                          NULL))
+        {
+          dnf_state_reset (hifstate);
+          g_autofree char *prefix = g_strdup_printf ("Updating metadata for '%s':",
```

Comment:
Ahh nice! I was toying with something similar a bit back. The default state steps in libdnf don't make any sense!